### PR TITLE
refactor(webbackend): Refactor exception's handler

### LINKED ISSUE #107: Refactor: Exception listing in Global Exception Handler
## What should be refactored?

- Move all type-defined exceptions definitions to dedicated providing class,
- Adjust switch statement to work with this class and provide details string to exception objects,

## Reason for refactoring:

Currently, in Global Exception Handler all possible exceptions are directly defined in switch statement's body.

### PR DESCRIPTION
**Pull request by Lukash88**
## Summary

Refactor global exception handler by adding static class GetProblemDetailsByExceptionFactory 
which returns appropriate ProblemDetails object according to incoming exception.

## Control List

- [x] Code formatting checked,
- [x] Coding Guidelines compliance checked,
- [ ] Documentation updated (if necessary),
- [ ] Automated tests updated (if necessary),

## Screenshots

Work under this PR is not visual, no screenshots available.

### CODE REVIEW

File: binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs [hunk 1]
```diff
@@ -11,6 +12,7 @@ public static IServiceCollection AddServices(this IServiceCollection services)
             services.AddScoped<IDefaultTableService, DefaultTableService>();
             services.AddScoped<IToDoTasksService, ToDoTasksService>();
             services.AddScoped<IAppVersionService, AppVersionService>();
+            services.AddScoped<IExceptionHandlerFactory, ExceptionHandlerFactory>();
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: There's no need in adding factory to IoC container - make it static class.

Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: Change class name to more descriptive about what is code doing there.

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 2]
```diff
@@ -1,61 +1,31 @@
...
-                        {
-                            Title = nameof(HttpStatusCode.InternalServerError),
-                            Status = (int)HttpStatusCode.InternalServerError,
-                            Detail = ExceptionConstants.UnexpectedErrorMessage,
-                        };
-                        context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
-                        break;
-                }
+                ProblemDetails? problemDetails = null;
+                _factory.HandleException(e, out problemDetails, context);
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Try to lower down amount of arguments for factory. Exception class only is sufficient to deliver all necessary data.

**Review by StraykerPL — CHANGES_REQUESTED** [review]

File: binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs (hunk 1 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: Resolved 64cca0b.

Comment (Lukash88) [reply binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: Resolved 64cca0b.

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 2 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Resolved 64cca0b.

File: binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 using Binder.Application.Models.Interfaces;
 using Binder.Application.Services;
+using Binder.Application.Services.Middleware;
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: Is this using required?

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 4]
```diff
@@ -1,68 +1,34 @@
...
-                        problemDetails = new ProblemDetails
-                        {
-                            Title = nameof(HttpStatusCode.InternalServerError),
-                            Status = (int)HttpStatusCode.InternalServerError,
-                            Detail = ExceptionConstants.UnexpectedErrorMessage,
-                        };
-                        context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
-                        break;
-                }
+                ProblemDetails? problemDetails = GetProblemDetailsByExceptionFactory.HandleException(e);
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Nullability not useful here.

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 5]
```diff
@@ -1,68 +1,34 @@
...
+                context.Response.StatusCode = (int)problemDetails.Status;
 
                 context.Response.ContentType = problemJsonType;
 
                 var json = JsonSerializer.Serialize(problemDetails);
-                await context.Response.WriteAsync(json);
+                await context.Response.WriteAsync(json);
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Delete spaces.

File: binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
+using System.Net;
+using Binder.Application.Services.Middleware.CustomExceptions;
+using Binder.Core.Constants;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Binder.Application.Services.Middleware
+{
+    public static class GetProblemDetailsByExceptionFactory
+    {
+        public static ProblemDetails HandleException(Exception exception)
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs]: Make method's name more descriptive, to point it directly to factory's meaning.

File: binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            ProblemDetails problemDetails;
+            switch (exception)
+            {
+                case InvalidOperationException:
+                    problemDetails = new ProblemDetails
```
Comment (StraykerPL) [inline binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs]: No need for variable here, save operation time with "return new" here.

**Review by StraykerPL — CHANGES_REQUESTED** [review]

File: binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs (hunk 3 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Extensions/ApplicationServices.cs]: Resolved 6cbce00.

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 4 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Resolved 6cbce00.

File: binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 5 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Resolved 6cbce00.

File: binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs (hunk 6 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs]: Resolved 6cbce00.

File: binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs (hunk 7 above)
Comment (Lukash88) [reply binder-web-backend/Binder.Application/Services/Middleware/GetProblemDetailsByExceptionFactory.cs]: Resolved 6cbce00.

**Review by StraykerPL — APPROVED** [review]

**Pull request merged**